### PR TITLE
Filter pods through pod list api

### PR DESCRIPTION
**Pull request by sujil02**
Refactored current filter pods flow through the shared pod functions
so filter pod functionalities can be shared between api and cmd.

Signed-off-by: Sujil02 <sushah@redhat.com>

### CODE REVIEW

File: cmd/podman/pod_ps.go [hunk 1]
```diff
@@ -174,49 +170,20 @@ func podPsCmd(c *cliconfig.PodPsValues) error {
...
-			generatedFunc, err := generatePodFilterFuncs(filterSplit[0], filterSplit[1])
-			if err != nil {
-				return errors.Wrapf(err, "invalid filter")
-			}
-			filterFuncs = append(filterFuncs, generatedFunc)
-		}
-	}
-
 	var pods []*adapter.Pod
-	if c.Latest {
```
Comment (mheon) [inline cmd/podman/pod_ps.go]: I think we need to retain the `c.Latest` code - that's not a filter, but a separate flag

File: cmd/podman/shared/pod.go [hunk 2]
```diff
@@ -134,4 +137,129 @@ func CreatePortBindings(ports []string) ([]ocicni.PortMapping, error) {
 	return portBindings, nil
 }
 
+func GetPodsWithFilters(r *libpod.Runtime, c *cliconfig.PodPsValues) ([]*libpod.Pod, error) {
+
+	var pods []*libpod.Pod
+	var err error
+	if c.Latest {
```
Comment (mheon) [inline cmd/podman/shared/pod.go]: I don't know if this should be part of it - @baude Thoughts?

File: cmd/podman/shared/pod.go [hunk 3]
```diff
@@ -134,4 +137,129 @@ func CreatePortBindings(ports []string) ([]ocicni.PortMapping, error) {
 	return portBindings, nil
 }
 
+func GetPodsWithFilters(r *libpod.Runtime, c *cliconfig.PodPsValues) ([]*libpod.Pod, error) {
```
Comment (mheon) [inline cmd/podman/shared/pod.go]: Comments on public functions are always appreciated

File: cmd/podman/shared/pod.go [hunk 4]
```diff
@@ -134,4 +137,129 @@ func CreatePortBindings(ports []string) ([]ocicni.PortMapping, error) {
...
+		pods, err = GenerateFilterFunctionAndGetPods(r, strings.Split(c.Filter, ","))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return pods, nil
+}
+
+func GenerateFilterFunctionAndGetPods(r *libpod.Runtime, filters []string) ([]*libpod.Pod, error) {
```
Comment (mheon) [inline cmd/podman/shared/pod.go]: Function comment would be appreciated

Comment (mheon) [reply cmd/podman/shared/pod.go]: Also, should this be public? Looks internal-only

File: cmd/podman/shared/pod.go [hunk 5]
```diff
@@ -134,4 +137,129 @@ func CreatePortBindings(ports []string) ([]ocicni.PortMapping, error) {
...
+		filterFuncs = append(filterFuncs, generatedFunc)
+	}
+	pods, err := r.Pods(filterFuncs...)
+	if err != nil {
+		return nil, err
+	}
+
+	return pods, nil
+}
+func GeneratePodFilterFuncs(filter, filterValue string) (
```
Comment (mheon) [inline cmd/podman/shared/pod.go]: As above - function comment, and should this be public?

File: pkg/bindings/test/pods_test.go [hunk 6]
```diff
@@ -49,7 +49,6 @@ var _ = Describe("Podman pods", func() {
 		Expect(err).To(BeNil())
 		Expect(response.Config.Name).To(Equal(newpod))
 	})
-
```
Comment (mheon) [inline pkg/bindings/test/pods_test.go]: Why was this removed?

File: pkg/bindings/test/pods_test.go [hunk 7]
```diff
@@ -77,14 +76,42 @@ var _ = Describe("Podman pods", func() {
...
 
-		// TODO not working  Because: code to list based on filter
-		// "not yet implemented",
-		// Validate list pod with filters
-		//filters := make(map[string][]string)
-		//filters["name"] = []string{newpod}
-		//filteredPods, err := pods.List(bt.conn, filters)
-		//Expect(err).To(BeNil())
-		//Expect(len(filteredPods)).To(BeNumerically("==", 1))
+		// Validate list pod with status filter
```
Comment (mheon) [inline pkg/bindings/test/pods_test.go]: Can these be put in a separate test?

**Comment by mheon** [comment]
@baude PTAL

File: cmd/podman/shared/pod.go (hunk 3 above)
Comment (sujil02) [reply cmd/podman/shared/pod.go]: Yeah. As mentioned I am working on the documentation just wanted to get the PR out to validate the code flow.

File: cmd/podman/shared/pod.go (hunk 2 above)
Comment (sujil02) [reply cmd/podman/shared/pod.go]: Should I put it back in pod_ps? So, Currently, it's not like that, but do we have to apply pods filter on the latest fetched pod as well if so we will need the flag here.

File: pkg/bindings/test/pods_test.go (hunk 7 above)
Comment (sujil02) [reply pkg/bindings/test/pods_test.go]: Done.

File: pkg/bindings/test/pods_test.go (hunk 6 above)
Comment (sujil02) [reply pkg/bindings/test/pods_test.go]: A typo maybe fixed now.

File: cmd/podman/shared/pod.go (hunk 5 above)
Comment (sujil02) [reply cmd/podman/shared/pod.go]: fixed. made private.

File: cmd/podman/shared/pod.go (hunk 4 above)
Comment (sujil02) [reply cmd/podman/shared/pod.go]: fixed. made private.

File: cmd/podman/shared/pod.go (hunk 3 above)
Comment (sujil02) [reply cmd/podman/shared/pod.go]: Added docs.

File: cmd/podman/shared/pod.go (hunk 2 above)
Comment (mheon) [reply cmd/podman/shared/pod.go]: Filters probably should not affect --latest - in fact, I'd go so far as to say filters and `--latest` should conflict.

File: libpod/runtime_pod.go [hunk 8]
```diff
@@ -90,18 +90,16 @@ func (r *Runtime) LookupPod(idOrName string) (*Pod, error) {
...
-		return nil, define.ErrRuntimeStopped
-	}
-
-	pods, err := r.state.AllPods()
+	pods, err := r.GetAllPods()
 	if err != nil {
 		return nil, err
 	}
 
+	return r.FilterPodsWithFilterFunc(pods, filters...)
```
Comment (mheon) [inline libpod/runtime_pod.go]: Why is this changed? I don't see any uses of FilterPodsWithFilterFunc. We also dropped the runtime validity check and lock?

Comment (sujil02) [reply libpod/runtime_pod.go]: Fixed now. I added this to support filters on the latest pod but this is no longer needed.  No, we did not drop that check there or was already a method [GetAllPods](https://github.com/containers/libpod/blob/master/libpod/runtime_pod.go#L121) that does the same just using that.

File: cmd/podman/pod_ps.go (hunk 1 above)
Comment (sujil02) [reply cmd/podman/pod_ps.go]: Fixed. The flag is retained.

**Comment by mheon** [comment]
/approve
LGTM
@baude @vrothberg @TomSweeneyRedHat @rhatdan PTAL

**Comment by TomSweeneyRedHat** [comment]
LGTM, nice work @sujil02!
